src/main/java/picard/sam/markduplicates/CheckDuplicateMarking.java

```diff
@@ -0,0 +1,173 @@
+        IOUtil.assertFileIsReadable(INPUT);
+        if (OUTPUT != null) {
+            IOUtil.assertFileIsWritable(OUTPUT);
+        }
+
+        try (SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT)) {
+
+            checkDuplicateMarkingsInIterable(getSortedRecordsFromReader(reader));
+            if (numBadRecords > 0) {
+                log.error("Found " + numBadRecords + " records that do not agree on their duplicate flag.");
+            } else {
+                log.info("All records' duplicate markings agree.");
+            }
+
+        } catch (IOException e) {
+            throw new PicardException("Error while reading input file " + INPUT, e);
+        }
+
+        return numBadRecords > 0 ? 1 : 0;
+    }
+
+    private Iterable<SAMRecord> getSortedRecordsFromReader(final SamReader reader) {
+        if (reader.getFileHeader().getSortOrder() == SAMFileHeader.SortOrder.queryname) {
+            return reader;
+        }
+
+        log.info("Input file isn't queryname sorted. Sorting into temp space.");
+
+        final Path[] tmpPaths = TMP_DIR.stream()
+                .map(File::toPath)
+                .toArray(Path[]::new);
+
+        final ProgressLogger sortProgress = new ProgressLogger(log, (int) 1e6, "Read into sorter");
+        final SortingCollection<SAMRecord> alignmentSorter = SortingCollection.newInstance(SAMRecord.class,
+                new BAMRecordCodec(reader.getFileHeader()),
+                new SAMRecordQueryNameComparator(),
+                MAX_RECORDS_IN_RAM,
+                tmpPaths);
+
+        for (final SAMRecord rec : reader) {
+            alignmentSorter.add(rec);
+            sortProgress.record(rec);
+        }
+        return alignmentSorter;
+    }
+
+    private boolean checkAndTallyRecordDuplicateMarking(final SAMRecord rec) {
+        if (!rec.getReadName().equals(currentReadName)) {
+            // this case the queryname changed, and thus there's no comparison to make
+            currentReadName = rec.getReadName();
+            currentReadDuplicateMarked = rec.getDuplicateReadFlag();
+        } else if (rec.getDuplicateReadFlag() != currentReadDuplicateMarked) {
+            // Here the current queryname is the same, but the duplicate flag doesn't match the first record with that queryname
+            numBadRecords++;
+
+            if (numBadRecords <= NUM_WARNINGS) {
+                log.warn(() -> "Reads with queryname " + currentReadName + " have different duplicate flags (at " +
+                        rec.getContig() + ":" + rec.getStart() + ")");
+            }
+
+            if (numBadRecords == NUM_WARNINGS) {
+                log.warn("Further warnings will be suppressed.");
+            }
+
+            return false;
+        }
+        return true;
+    }
+
+    private void checkDuplicateMarkingsInIterable(final Iterable<SAMRecord> iterable) throws IOException {
+        try (PrintWriter writer = OUTPUT == null ?
```

Comment:
could put ternary op inside constructor
```
        try (PrintWriter writer = new PrintWriter(OUTPUT != null ? new FileWriter(OUTPUT) : NullOutputStream.NULL_OUTPUT_STREAM)) {
```

Reply:
no...it's a different constructor..

Reply:
`PrintWriter(Writer)` vs.  `PrintWriter(OutputStream)`